src/test/java/com/ita/if103java/ims/service/WarehouseServiceImplTest.java

```diff
@@ -0,0 +1,454 @@
+        assertTrue(dBWarehouse.isActive());
+        when(addressDtoMapper.toEntity(warehouseDto.getAddressDto())).thenReturn(address);
+        Event event = new Event("Warehouse edited " +
+            updatedWarehouse.getName(), updatedWarehouse.getAccountID(),
+            updatedWarehouse.getId(), 1L, EventName.WAREHOUSE_EDITED, 2L);
+        doNothing().when(eventService).create(event);
+        when(warehouseDao.update(dBWarehouse)).thenReturn(updatedWarehouse);
+        assertFalse(updatedWarehouse.isTopLevel());
+
+        when(warehouseDao.findLevelByParentID(dBWarehouse.getParentID())).thenReturn(2);
+      when(warehouseDao.findByTopWarehouseID(4L, 1L)).thenReturn(List.of(new Warehouse()));
+        assertEquals(warehouseService.update(warehouseDto, userDetails), updatedWarehouse);
+    }
+
+    @Test
+    void update_notActive() {
+        Warehouse inActive = new Warehouse(12L, "WarehouseTest", "auto parts", 20, true, 5L, 1L, 4L, false);
+        when(warehouseDtoMapper.toEntity(warehouseDto)).thenReturn(inActive);
+        assertFalse(warehouse.isActive());
+        WarehouseUpdateException exception = assertThrows(WarehouseUpdateException.class, () -> {
+            warehouseService.update(warehouseDto, userDetails);
+        });
+        assertEquals("You can't make warehouse inactive!", exception.getMessage());
+    }
+
+    @Test
+    void update_notAllowChangeParent() {
+        Warehouse updatedWarehouse = new Warehouse(12L, "updatedWarehouse", "auto parts", 20, true, 5L, 1L, 4L, true);
+        Warehouse dBWarehouse = new Warehouse(12L, "dBWarehouse", "auto parts", 20, true, 5L, 1L, 4L, true);
+        warehouseDto = new WarehouseDto(12L, "WarehouseTest", "auto parts", 20, true, 5L, 1L, 4L, true, null);
+
+        when(warehouseDtoMapper.toEntity(warehouseDto)).thenReturn(updatedWarehouse);
+        when(warehouseDao.findById(updatedWarehouse.getId(), updatedWarehouse.getAccountID())).thenReturn(dBWarehouse);
+        warehouseDto.setId(3L);
+        assertTrue(updatedWarehouse.isActive());
+        updatedWarehouse.setActive(true);
+        if (!updatedWarehouse.getParentID().equals(dBWarehouse.getParentID())) {
+            WarehouseUpdateException exception = assertThrows(WarehouseUpdateException.class, () -> {
+                warehouseService.update(warehouseDto, userDetails);
+            });
+            assertEquals("You can't change parent warehouse!", exception.getMessage());
+        }
+    }
+
+
+    @Test
+    void findWarehousesByTopLevelId_Test() {
+        List<Warehouse> warehouseList = Arrays.asList(
+            new Warehouse(4L, "Store2", "auto parts", 0, false, 2L, 2L,
+                1L, true),
+            new Warehouse(3L, "Store3!", "tyres", 50, true, 2L, 2L,
+                1L, true));
+        when(warehouseDao.findByTopWarehouseID(1L, 1L)).thenReturn(List.of(warehouse));
+        when(warehouseDtoMapper.toDtoList(warehouseList)).thenReturn(List.of(warehouseDto));
+    }
```

Comment:
this test test nothing, there is no invocation of the warehouseService
Fix or remove

Reply:
This test still useless